### PR TITLE
Feature/itemlink ra

### PR DESCRIPTION
**Pull request by cydh**
* **Addressed Issue(s)**: -

* **Server Mode**: both

* **Description of Pull Request**: Just to generates `<ITEML>` includes base62 encode & decode

### CODE REVIEW

**Comment by Atemo** [comment]
How about including the feature in the atcommands `iteminfo`, `@mi` etc.. with config on/off eventually

File: src/common/utilities.hpp [hunk 1]
```diff
@@ -263,6 +263,24 @@ namespace rathena {
 		template <typename T> void tolower( T& string ){
 			std::transform( string.begin(), string.end(), string.begin(), ::tolower );
 		}
+
+		/**
+		* Pad string with arbitrary character in-place
+		* @param str String to pad
+		* @param padding Padding character
+		* @param num Maximum length of padding
```
Comment (aleos89) [inline src/common/utilities.hpp:271]:
```suggestion
		* @param str: String to pad
		* @param padding: Padding character
		* @param num: Maximum length of padding
```

File: src/common/utilities.hpp [hunk 2]
```diff
@@ -263,6 +263,24 @@ namespace rathena {
...
+		*/
+		void string_left_pad_inplace(std::string& str, char padding, int num);
+
+		/**
+		* Pad string with arbitrary character
+		* @param original String to pad
+		* @param padding Padding character
+		* @param num Maximum length of padding
```
Comment (aleos89) [inline src/common/utilities.hpp:279]:
```suggestion
		* @param original: String to pad
		* @param padding: Padding character
		* @param num: Maximum length of padding
```

File: src/map/itemdb.cpp [hunk 3]
```diff
@@ -2971,6 +2973,127 @@ void itemdb_reload(void) {
...
+	{ 'o' , 24 },{ 'p' , 25 },{ 'q' , 26 },{ 'r' , 27 },{ 's' , 28 },{ 't' , 29 },{ 'u' , 30 },{ 'v' , 31 },
+	{ 'w' , 32 },{ 'x' , 33 },{ 'y' , 34 },{ 'z' , 35 },{ 'A' , 36 },{ 'B' , 37 },{ 'C' , 38 },{ 'D' , 39 },
+	{ 'E' , 40 },{ 'F' , 41 },{ 'G' , 42 },{ 'H' , 43 },{ 'I' , 44 },{ 'J' , 45 },{ 'K' , 46 },{ 'L' , 47 },
+	{ 'M' , 48 },{ 'N' , 49 },{ 'O' , 50 },{ 'P' , 51 },{ 'Q' , 52 },{ 'R' , 53 },{ 'S' , 54 },{ 'T' , 55 },
+	{ 'U' , 56 },{ 'V' , 57 },{ 'W' , 58 },{ 'X' , 59 },{ 'Y' , 60 },{ 'Z' , 61 },
+};
+
+/**
+* Encode base10 number to base62. Originally by lututui
+* @param val Base10 Number
```
Comment (aleos89) [inline src/map/itemdb.cpp:3000]:
```suggestion
* @param val: Base10 Number
```

File: src/map/itemdb.cpp [hunk 4]
```diff
@@ -2971,6 +2973,127 @@ void itemdb_reload(void) {
...
+	while (val != 0) {
+		result = base62_dictionary[(val % 62)] + result;
+		val /= 62;
+	}
+	return result;
+}
+
+/**
+* Decode base62 string to base10. Originally by lututui
+* @param str Base62 String
```
Comment (aleos89) [inline src/map/itemdb.cpp:3018]:
```suggestion
* @param str: Base62 String
```

File: src/map/itemdb.cpp [hunk 5]
```diff
@@ -2971,6 +2973,127 @@ void itemdb_reload(void) {
...
+	}
+	catch (const std::out_of_range&) {
+		base10 = 0;
+	}
+	return base10;
+}
+
+/**
+* Generate <ITEML> string
+* @param data Item info
```
Comment (aleos89) [inline src/map/itemdb.cpp:3037]:
```suggestion
* @param data: Item info
```

File: src/map/script.cpp [hunk 6]
```diff
@@ -25677,6 +25677,78 @@ BUILDIN_FUNC( openstylist ){
...
+* Generate <ITEML> string for client
+* itemlink(<item_id>{,<refine>,<card0>,<card1>,<card2>,<card3>,<enchantgrade>});
+* itemlink2(<item_id>,<refine>,<card0>,<card1>,<card2>,<card3>,<enchantgrade>,<RandomIDArray>,<RandomValueArray>,<RandomParamArray>);
+* @author [Cydh]
+**/
+BUILDIN_FUNC(itemlink)
+{
+	struct s_item_link itemldata;
+
+	memset(&itemldata, 0, sizeof(s_item_link));
```
Comment (aleos89) [inline src/map/script.cpp:25690]:
```suggestion
	s_item_link itemldata = {};
```
No more memset, please! :(

File: src/map/script.cpp [hunk 7]
```diff
@@ -25677,6 +25677,78 @@ BUILDIN_FUNC( openstylist ){
...
+	FETCH(7, itemldata.item.card[3]);
+	FETCH(8, itemldata.item.enchantgrade);
+
+	if (itemldata.item.card[0] || itemldata.item.card[1] || itemldata.item.card[2] || itemldata.item.card[3])
+		itemldata.flag.cards = 1;
+
+#if PACKETVER >= 20150225
+	char* command = (char*)script_getfuncname(st);
+	if (command[strlen(command) - 1] == '2') { // only run when called via itemlink2
+		script_getitem_randomoption(st, NULL, &itemldata.item, command, 9);
```
Comment (aleos89) [inline src/map/script.cpp:25706]:
```suggestion
		script_getitem_randomoption(st, nullptr, &itemldata.item, command, 9);
```

**Review by aleos89 — COMMENTED** [review]
Looks good overall!

**Comment by Lemongrass3110** [comment]
Continued in #7291

**Pull request closed without merging**